File: src/rules/ConditionEmotionNode.cpp
#include <QtCore/QString>
#include <QtCore/QListIterator>

#include "ConditionEmotionNode.h"
#include "../Brain.h"
#include "../Emotion.h"

namespace CnotiMind
{

	ConditionEmotionNode::ConditionEmotionNode(const QString& emotion, const QString& value, ConditionNode::ConditionOperator op, Brain* brain, QObject* parent):
		ConditionNode( emotion, value, op, brain, parent ),
		_valueInt( value.toInt() )
	{

	}

	bool ConditionEmotionNode::isTrue() const
	{
		// If the brain doesn't have emotions
		if( _brain->_emotions.isEmpty() )
		{
			return false;
		}

		// search the emotion
		QListIterator<Emotion> it(_brain->_emotions);
		while( it.hasNext() )
		{
			Emotion e = it.next();

			if( e.key() == _key) // Emotion found
			{
				return testEmotion(e);
			}
		}

		return true;
	}

	/*
		Test the emotional value using the node operator
	*/
	bool ConditionEmotionNode::testEmotion( const Emotion& e ) const
	{
		switch( _operator )
		{
			case ConditionOperatorBigger: return e.value() == _valueInt;
			case ConditionOperatorBiggerOrEqual: return e.value() >= _valueInt;
			case ConditionOperatorSmaller: return e.value() < _valueInt;
			case ConditionOperatorSmallerOrEqual: return e.value() <= _valueInt;
			case ConditionOperatorEqual: return e.value() == _valueInt;
			case ConditionOperatorDifferent: return e.value() != _valueInt;
		}

		// if the operator is invalid
		return false;
	}

}

File: src_ios/CnotiMind/rules/ConditionEmotionNode.m
//
//  ConditionEmotionNode.m
//  CnotiMind
//
//  Created by Gonçalo Rodrigues on 11/05/09.
//  Copyright 2011 Universidade de Coimbra. All rights reserved.
//

#import "ConditionEmotionNode.h"

#import "Brain.h"

@implementation ConditionEmotionNode


- (id) initWithEmotionAndValueAndOperatorAndBrainAndParent: (NSString*)aEmotion 
                                                     value:(NSString*)aValue 
                                                  operator:(enum ConditionOperator)aOperator 
                                                     brain:(Brain*)aBrain 
                                                    parent:(id)aParent
{
    if (self == [super initWithKeyAndValueAndBrainAndParent: aEmotion value:aValue operator:aOperator brain:aBrain parent:aParent])
    {
    }
    return self;
}


- (void) exec
{
    if ([self isTrue]) {
        NSMutableDictionary* variables = [[NSMutableDictionary alloc] init];
        [variables setObject:_value forKey:@"[Emotion.value]"];
        
        //  TODO
        [self execChildren:variables];
    }
}


- (void) exec:(NSMutableDictionary*)aVariables
{
    if ([self isTrue]) {
        [aVariables setObject:_value forKey:@"[Emotion.value]"];
        [self execChildren];
    }

}


- (NSString*) info:(int)aDepth
{
    NSString* info = [NSString stringWithFormat:@""];
    NSString* space = [self space:aDepth];
    
    info = [info stringByAppendingFormat:@"%@ Condition type=Emotion", space];
    info = [info stringByAppendingFormat:@"%@", [super info:aDepth]];
    
    return info;
}


- (BOOL) isTrue
{
    
    // If the brain doesn't have emotions
    if( [_brain.emotions count] == 0)
    {
        return false;
    }
    
    // search the emotion
    NSEnumerator* eEmotions = [_brain.emotions objectEnumerator];
    Emotion* objectEmotion;
    while (objectEmotion = [eEmotions nextObject]) {
        if (![[objectEmotion name] caseInsensitiveCompare: _key]) {
            return [self testEmotion:objectEmotion];
        }
    }
    
    return true;
}


- (BOOL) testEmotion:(Emotion*)aEmotion
{
    if( _isValueNumeric )
    {
        switch( (int)_operator )
        {
            case ConditionOperatorBigger: return [aEmotion value] > _valueNumeric;
            case ConditionOperatorBiggerOrEqual: return [aEmotion value] >= _valueNumeric;
            case ConditionOperatorSmaller: return [aEmotion value] < _valueNumeric;
            case ConditionOperatorSmallerOrEqual: return [aEmotion value] <= _valueNumeric;
            case ConditionOperatorEqual: return [aEmotion value] == _valueNumeric;
            case ConditionOperatorDifferent: return [aEmotion value] != _valueNumeric;
        }
    }
    // if the operator is invalid or the value is not numeric
    return false;
}

@end
